### ISSUE TITLE
Dealing with jobs failing with "lost communication with the server" errors

### ISSUE THREAD

**Issue by mumoshu** [summerwind/actions-runner-controller#466]
I have rarely encountered this myself, but I believe any self-hosted GitHub runners are subject to get that error due to the race condition between the runner agent and GitHub.

Please see the related issues for more information.

- https://github.com/actions/runner/issues/510
- https://github.com/microsoft/azure-pipelines-agent/issues/2261
- https://github.com/apache/airflow/issues/14337
- https://github.com/actions/runner/issues/921#issuecomment-821118769

This issue is mainly to gather experiences from whoever has been affected by the error. I appreciate it if you could share your stories, workarounds, fixes, etc. around the issue so that it would ideally be fixed upstream or in actions-runner-controller.

Note that the error can also happen when:

- The runner container got OOM-killed due to that your runner pod has insufficient resource. Set higher resource requests/limits.
- The runner container got OOM-killed due to that your node has insufficient resource and your runner pod had low priority. Use a more resourceful machine as your node.

If you encounter the error even after tweaking your pod and node resources, it is likely that it's due to the race between the runner agent and GitHub.

Also note that disabling ephemeral runners (#457) (i.e. removing the `--once` flag from `run.sh`) may "alleviate" this issue, but not completely.

**Comment by jwalters-gpsw** [comment]
I'm still [dealing](https://github.com/summerwind/actions-runner-controller/issues/458) with the credentials not being set when the runner. But I could easily hack around the situation if I could set a liveness probe (exec) n my RunnerDeployment. There is both characteristic text in the log file as well as missing files in the /runner directory. Is that possible?

**Comment by mumoshu** [comment]
@jwalters-gpsw Hey! Stil trying to understand your issue, but isn't that different from what we're talking here? Are you seeing `lost communication with the server` message?

**Comment by jwalters-gpsw** [comment]
> @jwalters-gpsw Hey! Stil trying to understand your issue, but isn't that different from what we're talking here? Are you seeing `lost communication with the server` message?

Yes, the underlying issue is different. But if there is a way to use a liveness probe there might be a hacky solution to both problems by restarting the containers/pods/

**Comment by shreyasGit** [comment]
is  this issue resolved now? we are still getting this for self-hosted runners on Linux box?

**Comment by DerekTBrown** [comment]
> Note that the error can also happen when: The runner container got OOM-killed due to that your runner pod has insufficient resource. Set higher resource requests/limits.

Are there ways that `actions-runner-controller` or `actions-runner` could more gracefully handle the OOM-killed case? Could we somehow report OOM kills to the GitHub UI? Could we run a separate OOM killer inside the workers to kill the workflow before it exceeds memory limits?

**Comment by mumoshu** [comment]
@DerekTBrown Hey. I think that's fundamentally impossible. Even if we were able to hook into the host's system log to know about which process got OOM killed, we have no easy way to correlate it with a process within a container in a pod and even worse, GitHub Actions doesn't provide an API to "externally" set a workflow job status.

However, I thought you could still see a job times out after 10 minutes or so and the job that was running when the runner disappeared was eventually marked as failed(although without an explicit error message at all). Would that be enough?

**Comment by alyssaruth** [comment]
> This issue is mainly to gather experiences from whoever has been affected by the error. I appreciate it if you could share your stories, workarounds, fixes, etc. around the issue so that it would ideally be fixed upstream or in actions-runner-controller.

Really appreciate this! :heart: 

We currently see this occasionally - from our stats it's ~2.5% of builds of our main CI/CD pipeline. Our setup is that we're using `actions-runner-controller` within GKE, and the nodes that the runners get spun up on are [Spot VMs](https://cloud.google.com/kubernetes-engine/docs/concepts/spot-vms). We do this because it's a massive cost saving - however, it inevitably means sometimes we bump into this error when a node that's hosting a runner gets taken away from us by Google. 

Even outside of Spot VMs, there are all sorts of other imaginable reasons that are hard or impossible to mitigate - e.g. node upgrades, OOM kills and suchlike. The dream for us would be for jobs affected by this to automatically restart from the beginning, provisioning a fresh runner and going again.

**Comment by nsheaps** [comment]
@alyssa-glean for re-running, I'd recommend a workflow that runs on completion of the specific workflows you want (can use a cloud hosted or self-hosted runner for this) which re-triggers the job. We've been doing that at my place-of-work and it works great

As for `lost communication` errors, for us, this was mainly caused by custom timeout logic within our github actions workflows, which attempted to disable job control, then use the process group ID to determine which process to kill when the timeout expires. We've since changed this to using the linux native `timeout` command with job control and the problem has mostly resolved itself, except for some cases when github itself is having issues.

As for `runner shutdown` errors, this has been fully mitigated for us by the graceful termination suggestions [here](https://github.com/actions/actions-runner-controller/blob/master/docs/automatically-scaling-runners.md#configuring-automatic-termination). This has been confirmed by both the logs on the runner (which we export to ensure we still have them after the pod dies, by building a custom image that has fluentd in it which pipes the runner logs, runner worker logs, and the actions runner daemon to stdout), and prometheus metrics, which we send to datadog.

**What didn't work** is setting the pod label `cluster-autoscaler.kubernetes.io/safe-to-evict=false`, it seemed to have no effect (at least on EKS)

As for other general termination behavior, we've noticed our longer running jobs that utilize docker in docker/dind as a sidecar do not terminate gracefully. The system logs and metrics do note that a termination signal is sent, and the runner pod successfully waits. However, the default behavior of the container runtime/kubernetes is to send the termination signal to the main process on _all_ containers in a pod, including docker-in-docker. This kills the docker daemon and causes the tests to fail when the autoscaler decides that the pod should be terminated (which we haven't really figured out why it's happening in the first place). For us this equates to failures on ~13% of our job runs for these tests. It'd be nice for the sidecar to only get terminated if the runner itself is also exited, similar to the idea [here](https://carlosbecker.com/posts/k8s-sidecar-shutdown/). For now, my best idea (without adding more custom stuff to the image) would be to include dind in the runner container, rather than as a sidecar.

EDIT: The dind container DOES have that termination mechanism but the docs don't suggest how to set it properly. I peeked in the CRD and found the right way to set it (`dockerEnv`)

**Comment by signor-mike** [comment]
I downgraded Ubuntu from 22.04 LTS to 20.04 LTS and the workflow is no longer exhausting anything.

**Comment by bwodecki-lf** [comment]
Hi,

Do we have any known workaround for this issue?
Sometimes my actions get stuck in the running state for a while until they eventually time out. After that, I receive an error saying that the job lost communication with the runner.

The problem usually happens when the runner is restarted due to OOM issues or other errors.
Is there any way to reconnect an action to a runner after it has been restarted?

**Comment by Nikita-Guchakov** [comment]
We switched from gke to aws and same problems began.
ARC 13.x

First thing we found out that GKE internet gateway has longer idle tcp timeout, but AWS nat gateway has 350 secs. So we tweak KeepAlive settings ( send heart-beat faster ) It helped a bit, but still we still had a lot of Communication lost.

We had to increase our instances significantly, increased Listener priorities inside the container. 
Again, helped a bit, but still an issue.
Still an issue for us.

Idk were to look next, but ARC should have so much problems to be cooked properly at AWS. If there is so many caveats - doc should mention them and proper WA setup, if they are unavoidable. Though still seems like some race condition or another bug for me.
Our CPUs/memory seems to have enough space, someetimes it fails at steps that doesn't require cpu/memory at all

**Comment by Nikita-Guchakov** [comment]
For anyone who may have same problem: one of the biggest contributors here for us were AWS auto-scaling group a-z rebalancing enabled. You can check ASG logs to see if it rebalances nodes with the workers and disable it/deploy to single a-z

**Comment by Nikita-Guchakov** [comment]
As a solution from the side of the arc Pod distribution budget for runners can be a part of installation with minAvailable: 1, so any reason for node eviction will prevent pods from be killed
